manila/share/drivers/scality/driver.py

```diff
@@ -45,44 +72,157 @@
-        self._config_check()
-
-        self.export_ip = self.configuration.export_ip
-        self.ssh_pool = utils.SSHPool(
-            ip=self.configuration.export_management_host,
-            port=self.configuration.export_management_port,
-            conn_timeout=None,
-            login=self.configuration.management_user,
-            privatekey=self.configuration.ssh_key_path,
-        )
-
-    def _config_check(self):
-        """Validate configuration settings."""
-        log.debug('Validating configuration')
-        required = ('export_management_host', 'management_user', 'export_ip')
-        for config_item in required:
-            if getattr(self.configuration, config_item) is None:
-                msg = _("'%s' must be configured" % config_item)
-                raise exception.ManilaException(msg)
+
+        self._helpers = {}
+
+    def do_setup(self, context):
+        super(ScalityShareDriver, self).do_setup(context)
+
+        LOG.debug('Validating Scality Driver configuration')
+
+        msg = _("Configuration error: missing value for %(item)s. "
+                "Check the manila.conf file")
+
+        # Check generic configuration
+        for config_item in ('management_user', 'ssh_key_path'):
+            # Accept neither None nor ""
+            if not self.configuration.get(config_item):
+                raise exception.InvalidParameterValue(
+                    err=msg % {'item': config_item})
+
+        # Check protocol specific configuration
+        group_of_config_items = {
+            'NFS': ('nfs_export_ip', 'nfs_management_host'),
+            'SMB': ('smb_export_ip', 'smb_management_host',
+                    'smb_export_root')
+        }
+
+        for protocol, config_keys in group_of_config_items.items():
+            config_values = [self.configuration.get(k) for k in config_keys]
+
+            if not any(config_values):
+                # All the values are empty -> we don't want to configure
+                # this protocol
+                continue
+
+            if not all(config_values):
+                # Some values are missing
+                items = ' or '.join(config_keys)
+                raise exception.InvalidParameterValue(
+                    err=msg % {'item': items})
+
+            proto = protocol.lower()
+            ip = self.configuration.get(proto + '_management_host')
+            port = self.configuration.get(proto + '_management_port')
+            ssh_pool = utils.SSHPool(
+                ip=ip, port=port, conn_timeout=None,  # Unit test le timeout, sinon ca bloque indefiniement sur un firewall
+                login=self.configuration.management_user,
+                privatekey=self.configuration.ssh_key_path,
+                max_size=1)
+
+            export_ip = self.configuration.get(proto + '_export_ip')
+
+            if protocol == 'NFS':
+                self._helpers['NFS'] = NFSHelper(ssh_pool, export_ip)
+            else:
```

Comment:
Prone to error if ever we add some other protocol. The protocol not being NFS shouldn't imply it's SMB.